### PR TITLE
Add a module which implements spam protection via a maths problem.

### PR DESCRIPTION
**Pull request by SaberUK**

### CODE REVIEW

**Comment by md-5** [comment]
Even the silliest script kid can probably implement this into their bots.

**Comment by SaberUK** [comment]
In my experience 99% of spammers on IRC are kiddiots with no programming skill who have just downloaded a bot off the internet. It works against them.

**Comment by attilamolnar** [comment]
Exempting users who are on a channel is a bit easy to bypass, don't you think?

File: 2.0/m_solvemsg.cpp [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+		Problem problem;
+		problem.first = ServerInstance->GenRandomInt(9);
+		problem.second = ServerInstance->GenRandomInt(9);
+		problem.warned = false;
+		ext.set(user, problem);
+	}
+
+	ModResult OnUserPreMessage(User* user, void* dest, int target_type, std::string& text, char status, CUList& exempt_list)
+	{
+		if (user->exempt || target_type != TYPE_USER || !user->chans.empty())
```
Comment (attilamolnar) [inline 2.0/m_solvemsg.cpp]: `!user->chans.empty())` this is too broad IMO, should be configurable, what do you think?

Comment (SaberUK) [inline 2.0/m_solvemsg.cpp]: Sure. I'll add a config option for it.

**Comment by SaberUK** [comment]
@attilamolnar In the end I decided to just remove that check.

File: 2.0/m_solvemsg.cpp [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+			return MOD_RES_PASSTHRU;
+
+		Problem* problem = ext.get(user);
+		if (!problem)
+			return MOD_RES_PASSTHRU;
+
+		if (problem->warned)
+			return MOD_RES_DENY;
+
+		user->WriteServ("NOTICE %s :*** Before you can send messages you must join solve the following problem:", user->nick.c_str());
```
Comment (attilamolnar) [inline 2.0/m_solvemsg.cpp]: Typo/extra word here

File: 2.0/m_solvemsg.cpp [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+		if (!problem)
+		{
+			user->WriteServ("NOTICE %s :*** You have already solved your problem!", user->nick.c_str());
+			return CMD_FAILURE;
+		}
+
+		int result = ConvToInt(parameters[0]);
+		if (result != (problem->first + problem->second))
+		{
+			user->WriteServ("NOTICE %s :*** %s is not the correct answer.", user->nick.c_str(), parameters[0].c_str());
```
Comment (attilamolnar) [inline 2.0/m_solvemsg.cpp]: Raise the penalty to have a better protection against bruteforcing

**Comment by attilamolnar** [comment]
Thanks!

**Pull request merged**